### PR TITLE
remove inconsistency in cnode size calculations

### PR DESCRIPTION
**Pull request by malus-brandywine**
Capability Tutorial:

Since the first calculation gives the size of initial CNode in slots
the second calculation - division the number of slots by slot size -
has no meaning. The patch fixes the second calculation so it gives the
CNode size in bytes.

Signed-off-by: Nataliya Korovkina <malus.brandywine@gmail.com>

### CODE REVIEW

**Comment by pingerino** [comment]
Thanks for the contribution and suggestion. CNode sizing is indeed confusing. 

However, the tutorial aims to show you how to calculate both how much space the object takes, *and* how many slots. I don't think this change should be merged, but perhaps there is another change that could make this clearer?

**Comment by malus-brandywine** [comment]
What you say is what exactly the patch does. The final output will be:

Initial CNode is 65536 slots in size
The CNode is 2097152  bytes in size

(note: info->initThreadCNodeSizeBits = 16)

Like I has commented, the first calculation gives CNode size measured in slots. The code is (unchanged):

    size_t initial_cnode_object_size = BIT(info->initThreadCNodeSizeBits);
    printf("Initial CNode is %zu slots in size\n", initial_cnode_object_size);

The second calculation (patched) measures it in bytes. The code is:

    size_t initial_cnode_object_size_bytes = initial_cnode_object_size * (1u << seL4_SlotBits);
    printf("The CNode is %zu  bytes in size\n", initial_cnode_object_size_bytes);

In contrast to original one:

    size_t num_initial_cnode_slots = initial_cnode_object_size / (1u << seL4_SlotBits);

(note: initial_cnode_object_size is the size expressed in slots)

which has no meaning, because it divides number of slots by slot size.

Where am I wrong?

Thank you.

**Comment by lsf37** [comment]
I think you're right, the tutorial incorrectly assumed that `info->initThreadCNodeSizeBits` is the log of the size, but it is actually the log of the number of slots, and so the second calculation is wrong.

I'm happy with the PR, but I'd like to track down why the first reported value is different. I thought the tests check the output, and should have picked up when a size different from 4096 is reported. Will look into that.

**Comment by malus-brandywine** [comment]
It slipped from my attention that the clause "CNodes and CSlots" contains incorrectness as well.

Currently, the last paragraph is:

CSlots are 1u << seL4_SlotBits in size, and as a result the number of slots in a CNode can be calculated by CNodeSize / (1u << seL4_SlotBits).

</br>
I would suggest to replace it with:
</br></br>

Sizes of the objects. In general in seL4, if sizes are measured in bits (`size_bits`), the object size is 2<sup>`size_bits`</sup> bytes (see the [Untyped] tutorial). CNode is an exception, it is measured in CSlots, so the size is 2<sup>`size_bits`</sup> CSlots. CSlots are `seL4_SlotBits` in size.

[untyped]: https://docs.sel4.systems/Tutorials/untyped.html

</br>

I assume, that's enough to fulfil the task.
Grammar and style corrections are welcome.

Gerwin, when you figure out what the different number came from, could you let me know if any other changes should be done? I will compose all the changes in a single patch, make a new PR and will close current one.

Thank you!

**Comment by lsf37** [comment]
> CSlots are 1u << seL4_SlotBits in size, and as a result the number of slots in a CNode can be calculated by CNodeSize / (1u << seL4_SlotBits).

The sentence as such is not wrong, at least if `CNodeSize` really is the size (in bytes), but I agree that it is confusing in this context. Since other seL4 objects haven't been introduced yet, I'd prefer to avoid talking about these. How about the following instead?

> The field `info->CNodeSizeBits` gives a measure of the size of the initial CNode: it will have `1 << CNodeSizeBits` CSlots. A CSlot has `1 << seL4_SlotBits` bytes, so the size of a CNode in bytes is `1 << (CNodeSizeBits + seL4_SlotBits)`.

For the test:

> Gerwin, when you figure out what the different number came from, could you let me know if any other changes should be done? I will compose all the changes in a single patch, make a new PR and will close current one.

I had a look at the test, and it doesn't expect a specific number, which makes sense, because the number depends on what platform you run the tutorial on and how the kernel is configured for it. Are you getting the size 16 initial CNode for PC99?

In the meantime, I've enabled the test for pull-request changes to markdown files, so we can see the result directly here. The capability tutorial fails, because we have to adjust the expected output string as well as the code (will add a to the code directly).

You can add additional commits to this PR by pushing to your branch -- they will show up here, and when we have reached a state that we are all happy with, we can turn them into a single commit (no need to open a new PR).

File: tutorials/capabilities/capabilities.md [hunk 1]
```diff
@@ -224,29 +224,29 @@ The third line stating the number of slots in the CSpace, is incorrect, and your
...
-    printf("The CSpace has %zu CSlots\n", num_initial_cnode_slots);
+    size_t initial_cnode_object_size_bytes = 0; // TODO calculate this.
+    printf("The CNode is %zu bytes in size\n", initial_cnode_object_size_bytes);
 /*-- endfilter -*/
 ```
 
 /*-- filter ExcludeDocs() -*/
 
 ```c
 /*-- filter TaskContent("cnode-size", TaskContentType.COMPLETED, subtask='size', completion='The CSpace has [0-9]+ CSlots') -*/
```
Comment (lsf37) [inline tutorials/capabilities/capabilities.md:239]:
```suggestion
/*-- filter TaskContent("cnode-size", TaskContentType.COMPLETED, subtask='size', completion='The CNode is [0-9]+ bytes in size') -*/
```

**Comment by pingerino** [comment]
> What you say is what exactly the patch does. 

Ah sorry, I should stop trying to squeeze in a few PRs before coffee :)

**Comment by lsf37** [comment]
Ok, this looks good now. I'll squash the commits into one to make sure all the tests pass, and then we can merge.

**Comment by lsf37** [comment]
(removed trailing whitespace)

**Review by lsf37 — APPROVED** [review]

**Comment by malus-brandywine** [comment]
Gerwin, thanks for your help and patience!

**Comment by lsf37** [comment]
And thanks for not only spotting the problem, but also fixing it!

**Pull request merged**